src/cpp/shared/game/game.cpp

```diff
@@ -55,8 +56,32 @@ void game::RandomizeHeroPool() {
 	}
 }
 
+extern char gMapName[]; // the map selected in the list
+extern char gLastFilename[];
+extern signed char xIsExpansionMap;
+
+void game::InitNewGame(struct SMapHeader *a) {
+	if (!strlen(gLastFilename)) { // game just started, no map was played yet
+		std::string lastPlayed;
+		if (xIsExpansionMap)
+			lastPlayed = read_pref<std::string>("Last Map expansion");
+		else
+			lastPlayed = read_pref<std::string>("Last Map");
+
+		if (lastPlayed.length() < 20) { // otherwise means no registry keys exist yet
+			strcpy(gMapName, lastPlayed.c_str());
+			strcpy(this->mapFilename, lastPlayed.c_str());
+		}
+	}
+	this->InitNewGame_orig(a);
+}
+
 void game::NewMap(char* mapname) {
 	send_event(mapAction, mapname);
+	if (xIsExpansionMap)
+		write_pref("Last Map expansion", std::string(gMapName));
```

Comment:
We have some more registry related code somewhere but it's not using named constants. I didn't do this here for the sake of consistency. Probably this is not related to this branch.